notebooks/work-in-progress/ferc1-transform-refactor/ferc1_transform_refactor.py

```diff
@@ -0,0 +1,429 @@
+            "fuel_type": FUEL_TYPES,
+        },
+    },
+}
+
+
+################################################################################
+# Pydanic Transformation Parameter Models
+################################################################################
+class TransformParams(BaseModel):
+    """An immutable base model for transformation parameters."""
+
+    class Config:
+        """Prevent parameters from changing part way through."""
+
+        allow_mutation = False
+
+
+class MultiColumnTransformParams(TransformParams):
+    """Transform params that apply to several columns in a table.
+
+    The keys are column names, and the values must all be the same type of
+    :class:`TransformParams` object, since MultiColumnTransformParams are used by
+    :class:`MultiColumnTransformFn` callables.
+
+    Individual subclasses are dynamically generated for each multi-column transformation
+    specified within a :class:`TableTransformParams` object.
+
+    """
+
+    @root_validator
+    def single_param_type(cls, params):  # noqa: N805
+        """Check that all TransformParams in the dictionary are of the same type."""
+        param_types = {type(params[col]) for col in params}
+        if len(param_types) > 1:
+            raise ValueError(
+                "Found multiple parameter types in multi-column transform params: "
+                f"{param_types}"
+            )
+        return params
+
+
+class RenameColumns(TransformParams):
+    """A dictionary for mapping old column names to new column names in a dataframe."""
+
+    columns: dict[str, str] = {}
+
+
+class StringCategories(TransformParams):
+    """Defines mappings to clean up manually categorized freeform strings.
+
+    Each key in a stringmap is a cleaned output category, and each value is the set of
+    all strings which should be replaced with associated clean output category.
+
+    """
+
+    categories: dict[str, set[str]]
+
+    @property
+    def mapping(self) -> dict[str, str]:
+        """A 1-to-1 mapping appropriate for use with :meth:`pd.Series.map`."""
+        return {
+            string: cat for cat in self.categories for string in self.categories[cat]
+        }
+
+
+class UnitConversion(TransformParams):
+    """A column-wise unit conversion."""
+
+    multiplier: float = 1.0
+    adder: float = 0.0
+    pattern: typing.Pattern
+    repl: str
+
+
+class TableTransformParams(TransformParams):
+    """All defined transformation parameters for a table."""
+
+    class Config:
+        """Only allow the known table transform params."""
+
+        extra = "forbid"
+
+    rename_columns: RenameColumns = {}
+    convert_units: dict[str, UnitConversion] = {}
+    categorize_strings: dict[str, StringCategories] = {}
+
+    @classmethod
+    def from_id(cls, table_id: str) -> "TableTransformParams":
+        """A factory method that looks up transform parameters based on table_id."""
+        return cls(**TRANSFORM_PARAMS[table_id])
```

Comment:
Accessing a large dict stored within a package makes it difficult for other projects to reuse these classes. We have a similar issue with the metadata classes. Could the metadata be read from a file or database?

Reply:
I was using the `from_id()` method to provide some decoupling here, so we can read in this information from elsewhere later if we want. Pydantic is happy to read parameters from JSON as well. The types that are natively available in JSON are more limited than in Python, but I think we can get around that by converting the types when necessary in the Pydantic validators (e.g. turning a JSON `list` into a Python `set` in the string categorization parameters.

At the same time, I think it's nice to be able to see all the parameters enumerated here in the module that they pertain to, so if you're wondering what all is happening to a given table or or column, you can look at see pretty directly.

Reply:
Got it. How can design this method so these classes can be built from any `TRANSFORM_PARAMS` dict?

Reply:
We could store the relevant `TableTransformParams` inside the `TableTransformer` class itself, rather than looking it up from some external source, but I thought the idea was that to work smoothly with Dagster we need to be able to instantiate the class based just on the name of the table (rather than passing in a gigantic `config` dictionary).